mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/RequestState.scala

```diff
@@ -375,29 +401,53 @@ import scala.util.{Failure, Success}
 
   // Processing
 
-  def main(registrantsByPacketId: Map[PacketId, ActorRef[A]], nextPacketId: PacketId): Behavior[Request[A]] =
-    Behaviors.receiveMessage {
-      case Register(registrant: ActorRef[A], reply) if nextPacketId.underlying <= MaxPacketId.underlying =>
-        reply.success(Registered(nextPacketId))
-        main(registrantsByPacketId + (nextPacketId -> registrant), PacketId(nextPacketId.underlying + 1))
-      case _: Register[A] =>
-        Behaviors.same // We cannot allocate any more. This will eventually cause a timeout to occur on the requestor.
-      case Unregister(packetId) =>
-        val remainingPacketIds = registrantsByPacketId - packetId
-        val revisedNextPacketId = if (remainingPacketIds.nonEmpty) {
-          val maxPacketId = remainingPacketIds.keys.maxBy(_.underlying)
-          PacketId(maxPacketId.underlying + 1)
-        } else {
-          MinPacketId
-        }
-        main(remainingPacketIds, revisedNextPacketId)
-      case Route(packetId, event, failureReply) =>
-        registrantsByPacketId.get(packetId) match {
-          case Some(reply) => reply ! event
-          case None => failureReply.failure(CannotRoute(packetId))
-        }
-        Behaviors.same
-    }
+  def main(registrantsByPacketId: Map[PacketId, ActorRef[A]],
+           nextPacketId: Option[PacketId],
+           pendingRegistrations: Vector[Register[A]]): Behavior[Request[A]] =
+    Behaviors
+      .receive[Request[A]] {
+        case (_, register @ Register(registrant: ActorRef[A], reply)) =>
+          nextPacketId match {
+            case Some(currentPacketId) =>
+              reply.success(Registered(currentPacketId))
+
+              val nextRegistrations = registrantsByPacketId + (currentPacketId -> registrant)
+
+              main(
+                nextRegistrations,
+                findNextPacketId(nextRegistrations, currentPacketId),
+                pendingRegistrations
+              )
+
+            case None =>
+              // all packet ids are taken, so we'll wait until one is unregistered
+              // to continue
+
+              main(registrantsByPacketId, nextPacketId, pendingRegistrations :+ register)
```

Comment:
I'm still unsure as to the benefit of having a `pendingRegistrations` structure. It should be a rare condition and adds complexity.